sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpTransactionEnlistment.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+using System.Transactions;
+using Microsoft.Azure.Amqp;
+using Azure.Messaging.ServiceBus.Primitives;
+using Microsoft.Azure.Amqp.Transaction;
+
+namespace Azure.Messaging.ServiceBus.Amqp
+{
+    internal class AmqpTransactionEnlistment : Singleton<AmqpTransactionEnlistment>, IPromotableSinglePhaseNotification
+    {
+        private readonly string _transactionId;
+        private readonly AmqpTransactionManager _transactionManager;
+        private readonly AmqpConnectionScope _connectionScope;
+        private readonly TimeSpan _timeout;
+
+        public AmqpTransactionEnlistment(
+            Transaction transaction,
+            AmqpTransactionManager transactionManager,
+            AmqpConnectionScope connectionScope,
+            TimeSpan timeout)
+        {
+            _transactionId = transaction.TransactionInformation.LocalIdentifier;
+            _transactionManager = transactionManager;
+            _connectionScope = connectionScope;
+            _timeout = timeout;
+        }
+
+        public ArraySegment<byte> AmqpTransactionId { get; private set; }
+
+        protected override async Task<AmqpTransactionEnlistment> OnCreateAsync(TimeSpan timeout)
+        {
+            try
+            {
+                FaultTolerantAmqpObject<Controller> faultTolerantController = _connectionScope.TransactionController;
+                Controller controller = await faultTolerantController.GetOrCreateAsync(timeout).ConfigureAwait(false);
+                AmqpTransactionId = await controller.DeclareAsync().ConfigureAwait(false);
+                MessagingEventSource.Log.AmqpTransactionDeclared(_transactionId, AmqpTransactionId);
+                return this;
+            }
+            catch (Exception exception)
+            {
+                MessagingEventSource.Log.AmqpTransactionInitializeException(_transactionId, exception);
+                _transactionManager.RemoveEnlistment(_transactionId);
+                throw;
+            }
+        }
+
+        protected override void OnSafeClose(AmqpTransactionEnlistment value)
+        {
+        }
+
+        void IPromotableSinglePhaseNotification.Initialize()
```

Comment:
I can't because it is an explicitly implemented interface method. I can implement implicitly but I copied this from Track 1 and it seemed more clear to have it be explicit.